### PR TITLE
move ssl client setup to main context

### LINKED ISSUE #7: Pull out SSL client setup
Right now, we embed the SSL setup inside the TCP connection setup. This means we're running SSL setup per request, which sounds like unnecessary extra overhead. Only the TCP needs to be re-established, while we can probably keep a copy of the SSL initialization per PEM.

### PR DESCRIPTION
**Pull request by travisstaloch**
Addresses and possibly closes #7.  Also cleans up my last pr so that in that github auth is formatted only once.  

I'm not sure that the ssl_tunnel deinit strategy is 100% correct.  This pr required a change to `Https.deinit()`, removing its ` self.ssl_tunnel.deinit();` which prevents segfaults from happening (introduced by this pr).  I hope this doesn't mess with your recent updates related to `heartbeat_ack` and `is_dying` flags which call `ssl_tunnel.deinit()`.  

Let me know if you'd like me to change anything.

### CODE REVIEW

**Comment by travisstaloch** [comment]
i'm pretty sure there are actually problems with this pr.  i believe the ssl trust_anchor and x509 in `SslTunnel.init()` are being clobbered and need to be heap allocated now.  i originally had written this way but changed back to stack allocating them after i identified what was causing the segfault i mentioned above.

**Comment by travisstaloch** [comment]
Noticed two threads pegging and reran with gdb.  Got this output:
```sh
[LWP 30815 exited]

Thread 1 "zigbot9001" received signal SIGSEGV, Segmentation fault.
0x000000000026dd67 in request.SslTunnel.connect (self=0x7ffff7fda000, args=...) at request.zig:27
27	        self.client = ssl.Client.init(self.x509.getEngine());
(gdb) c
Continuing.
Segmentation fault at address 0x7ffff7fda010
~/Code/zig/zigbot9001/src/request.zig:27:48: 0x26dd67 in request.SslTunnel.connect (zigbot9001)
        self.client = ssl.Client.init(self.x509.getEngine());
                                               ^
~/Code/zig/zigbot9001/src/main.zig:476:38: 0x256c9d in DiscordWs.init (zigbot9001)
        try result.ssl_tunnel.connect(.{
                                     ^
~/Code/zig/zigbot9001/src/main.zig:343:44: 0x254d20 in main (zigbot9001)
        var discord_ws = try DiscordWs.init(context);
                                           ^
~/Code/zig/zig/build/lib/zig/std/start.zig:257:37: 0x2586de in std.start.main (zigbot9001)
            const result = root.main() catch |err| {
                                    ^
~/Code/zig/zig/build/lib/zig/libc/musl/src/env/__libc_start_main.c:94:7: 0x3c4dc4 in libc_start_main_stage2 (~/Code/zig/zig/build/lib/zig/libc/musl/src/env/__libc_start_main.c)
	exit(main(argc, argv, envp));
      ^
```

I'm wondering if I need to remove the `ssl_tunnel.deinit()`s from heartbeatHandler and `DiscordWs.deinit()`?  Trying this and running again.

**Comment by travisstaloch** [comment]
The segfaults I mentioned above were caused by the deinits which have been removed in this pr.  Trying to solve this by heap allocating the ssl trust anchor and x509 was a red herring.

I've been running the these changes for an hour and no issues.  Will share if anything happens.

**Comment by travisstaloch** [comment]
One thing i'd like to add: `defer context.deinit()` to main.  Also hide the 3 new `SslTunnel.init()`s within `Context.init()` to shorten up main. 

Going to hold off on making any more changes until reviewed though - incase I need to backtrack first.

File: src/request.zig [hunk 1]
```diff
@@ -19,41 +19,41 @@ pub const SslTunnel = struct {
 
     pub const Stream = ssl.Stream(*std.fs.File.Reader, *std.fs.File.Writer);
 
-    pub fn init(args: struct {
+    pub fn connect(self: *SslTunnel, args: struct {
```
Comment (fengb) [inline src/request.zig:22]: Can you move this and the TCP related fields into Https? Keeping the connection data separate would prevent some potential concurrency bugs.

File: src/main.zig [hunk 2]
```diff
@@ -633,7 +668,6 @@ const DiscordWs = struct {
 
             if (!self.heartbeat_ack) {
                 self.is_dying = true;
-                self.ssl_tunnel.deinit();
```
Comment (fengb) [inline src/main.zig:636]: The main reason this exists is to signal a termination the main (listening) thread. It's a bit blunt, but right now the disconnecting the connection will cause `wz.readEvent()` to trigger `error.IO`. I think we can get that behavior back by closing the raw TCP connection here.

File: src/request.zig (hunk 1 above)
Comment (travisstaloch) [reply src/request.zig:22]: sounds good. hopefully start on this today.

File: src/main.zig (hunk 2 above)
Comment (travisstaloch) [reply src/main.zig:636]: you're referring to both deinits() here correct?

File: src/request.zig [hunk 3]
```diff
@@ -94,14 +94,13 @@ pub const Https = struct {
...
             .allocator = args.allocator,
-            .ssl_tunnel = ssl_tunnel,
+            .ssl_tunnel = args.ssl_tunnel,
             .buffer = buffer,
             .client = client,
         };
     }
 
     pub fn deinit(self: *Https) void {
-        self.ssl_tunnel.deinit();
```
Comment (travisstaloch) [inline src/request.zig:104]: Did you notice this one too? Does this look ok?

Comment (fengb) [reply src/request.zig:104]: Oh, once the TCP connection moves here, we should be closing that here. But it makes sense to leave the SSL part alone.

File: src/main.zig (hunk 2 above)
Comment (travisstaloch) [reply src/main.zig:636]: Handing off here after discussion on discord.

**Comment by fengb** [comment]
Thanks for exploring this. With your efforts and the new iguana code, it’s more obvious which direction I’d like to go: https://github.com/fengb/zCord/issues/2

**Pull request closed without merging**